In [1]:
import pandas as pd
import numpy as np
import re

In [2]:
# đọc dữ liệu
df = pd.read_csv(r"D:\I. KhongPhanLoai\IT-job-analysis-VN\data_raw\ITViec_data.csv")

In [3]:
# xem thông tin cơ bản về bộ dữ liệu
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 993 entries, 0 to 992
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   job_names             993 non-null    object
 1   company_names         993 non-null    object
 2   salaries              993 non-null    object
 3   position_names        993 non-null    object
 4   kind_jobs             993 non-null    object
 5   array_skills          993 non-null    object
 6   locate_names          993 non-null    object
 7   exp_skills            993 non-null    object
 8   domain_arr            993 non-null    object
 9   post_dates_formatted  993 non-null    object
dtypes: object(10)
memory usage: 77.7+ KB


bộ dữ liệu bao gồm 993 mẫu, 10 thuộc tính bao gồm:
- job_names: tên công việc
- company_names: tên công ty
- salaries: mức lương
- kind_job: loại công việc (toàn thời gian, bán thời gian, thực tập, vv)
- array_skills: các công cụ sử dụng
- locate_names: địa điểm làm việc
- exp_skills: kinh nghiêm và kĩ năng yêu cầu
- domain_arr: lĩnh vực công việc
- post_dates_formatted: ngày đăng tuyển
==> tất cả là biến phân loại trừ salary.

# KIỂM TRA NULL, NAN, DUPLICATE

In [4]:
# kiểm tra giá trị null
df.isnull().sum()

job_names               0
company_names           0
salaries                0
position_names          0
kind_jobs               0
array_skills            0
locate_names            0
exp_skills              0
domain_arr              0
post_dates_formatted    0
dtype: int64

In [5]:
# kiểm tra dữ liệu trùng
df[df.duplicated(keep=False)]

,job_names,company_names,salaries,position_names,kind_jobs,array_skills,locate_names,exp_skills,domain_arr,post_dates_formatted
19,Automation Tester,Simpson Strong-Tie Vietnam,You'll love it,Automation Tester,Hybrid (flexible between home and office),"['Automation Test', 'C#', 'QA QC', 'JavaScript...","9th Floor, Etown 6 Building, 364 Cong Hoa Stre...",Desired skills and experience: \nBachelor's de...,['Software Products and Web Services'],25/11/2025 17:12
20,Automation Tester,Simpson Strong-Tie Vietnam,You'll love it,Automation Tester,Hybrid (flexible between home and office),"['Automation Test', 'C#', 'QA QC', 'JavaScript...","9th Floor, Etown 6 Building, 364 Cong Hoa Stre...",Desired skills and experience: \nBachelor's de...,['Software Products and Web Services'],25/11/2025 17:12
26,Manual Tester (QA QC),Công ty cổ phần Tập đoàn Công nghệ Quảng Ích (...,13-20m,Manual Tester,At office,"['Tester', 'Mobile Apps', 'QA QC']","46 LePARC by Gamuda, Công Viên Yên Sở, Hoang M...",Có từ 2 năm trở lên kinh nghiệm Tester.\nNhanh...,['Education and Training'],25/11/2025 16:12
27,Manual Tester (QA QC),Công ty cổ phần Tập đoàn Công nghệ Quảng Ích (...,13-20m,Manual Tester,At office,"['Tester', 'Mobile Apps', 'QA QC']","46 LePARC by Gamuda, Công Viên Yên Sở, Hoang M...",Có từ 2 năm trở lên kinh nghiệm Tester.\nNhanh...,['Education and Training'],25/11/2025 16:12
57,"Senior Software Developer (C#, .Net, SQL)",MiTek Viet Nam,You'll love it,Fullstack Developer,At office,"['.NET', 'English', 'Azure', 'SQL', 'WinForms'...","Tòa nhà A5, Lô số A5, khu E-Office, đường Sáng...",Required Qualifications:\nBachelor of Science ...,['Software Products and Web Services'],25/11/2025 09:13
58,"Senior Software Developer (C#, .Net, SQL)",MiTek Viet Nam,You'll love it,Fullstack Developer,At office,"['.NET', 'English', 'Azure', 'SQL', 'WinForms'...","Tòa nhà A5, Lô số A5, khu E-Office, đường Sáng...",Required Qualifications:\nBachelor of Science ...,['Software Products and Web Services'],25/11/2025 09:13
60,Java Developer (3 years +),CODE88 COMPANY LIMITED,"1,000 - 2,500 USD",Backend Developer,At office,"['Spring', 'ReactJS', 'Spring Boot']","6th Floor, Menas Mall Saigon Airport, 60A Truo...",Skills & Qualifications:\nSoftware development...,['Financial Services'],25/11/2025 07:14
61,Java Developer (3 years +),CODE88 COMPANY LIMITED,"1,000 - 2,500 USD",Backend Developer,At office,"['Spring', 'ReactJS', 'Spring Boot']","6th Floor, Menas Mall Saigon Airport, 60A Truo...",Skills & Qualifications:\nSoftware development...,['Financial Services'],25/11/2025 07:14
63,Principal QA QC Engineer (Manual & Automation ...,"MONEY FORWARD VIETNAM CO.,LTD",You'll love it,Automation Tester,At office,"['QA QC', 'LLM', 'API', 'AI', 'A/B testing', '...","ETown Central Building, 11 Doan Van Bo, Ward 1...",Technical\nMust have\n5+ years of experience i...,"['Software Products and Web Services', 'Financ...",25/11/2025 00:14
64,Principal QA QC Engineer (Manual & Automation ...,"MONEY FORWARD VIETNAM CO.,LTD",You'll love it,Automation Tester,At office,"['QA QC', 'LLM', 'API', 'AI', 'A/B testing', '...","ETown Central Building, 11 Doan Van Bo, Ward 1...",Technical\nMust have\n5+ years of experience i...,"['Software Products and Web Services', 'Financ...",25/11/2025 00:14


In [6]:
# xóa dữ liệu trùng
df = df.drop_duplicates()

# CLEAN CỘT JOB_NAMES

In [7]:
# kiểm tra số lượng các job tương tự nhau
df.job_names.nunique()
# ==> có 961 job khác nhau ==> vô lý

961

# đặc điểm của cột job_names:
- tên job bao gồm tên job, cấp bậc, công ty tuyển dụng, yêu cầu ngôn ngữ, công cụ sử dụng, loại hình công việc
- có nhiều job trùng tên nhưng có thêm loại hình làm việc
- có nhiều job trùng tên nhưng khác công ty tuyển dụng
- có nhiều job trùng tên nhưng khác yêu cầu ngôn ngữ, công cụ sử dụng
- có nhiều job trùng tên nhưng khác cấp bậc
- thậm chí có nhiều job không có tên công việc rõ ràng

* hướng giải quyết:
- giữ tên job chính, các thông tin về level, yêu cầu ngôn như, công cụ sử dụng hay loại hình công việc sẽ giữ lại để tiếp tục mổ xẻ để lấy thông tin tiếp
- các bước:
+ lấy job chính (thường là chỉ có tên job, không nằm trong ngoặc, không có dấu gạch ngang, không có dấu phẩy)
+ lọc và loại các yêu cầu level
* không được xóa hết thông tin mà chỉ đổ sang cột khác, nếu không có thì có thể cho null để tránh sai lệch dữ liệu

In [8]:
# lọc sửa clean lần 1

levels = ["intern","fresher","all level", "junior", "mid", "middle", "senior", "lead", "principal"]
pattern_level = r'(' + '|'.join(levels) + r')'

# 1. Tách level (không phân biệt hoa thường)
df["level"] = df["job_names"].str.lower().str.extract(pattern_level, expand=False)

# 2. Tách nội dung trong ngoặc → xu_ly_sau
df["xu_ly_sau"] = df["job_names"].str.extract(r'\((.*?)\)', expand=False)

# 3. Tạo cột job_names_clean
def clean_job_name(job):
    if pd.isna(job):
        return None
    # Loại bỏ dấu ngoặc và nội dung trong ngoặc
    job = re.sub(r'\(.*?\)', '', job)
    # Loại bỏ dấu gạch ngang và dấu phẩy
    job = re.sub(r'[-,]', '', job)
    # Loại bỏ level đã tách
    for level in levels:
        pattern = re.compile(level, flags=re.IGNORECASE)
        job = pattern.sub('', job)
    # Xóa khoảng trắng thừa ở đầu cuối
    job = job.strip()
    return job

df["job_names_clean"] = df["job_names"].apply(clean_job_name)

# Giữ nguyên index gốc
df = df.sort_index()


In [9]:
# lọc sửa clean lần 2

def clean_job_name_step4(job):
    if pd.isna(job):
        return pd.Series([None, None, None])  # ky_tu, xu_ly_sau2, job_names_clean2
    
    ky_tu = []
    xu_ly_sau2 = []

    # 1. Lấy nội dung trong [...] sang xu_ly_sau2
    bracket_matches = re.findall(r'\[(.*?)\]', job)
    if bracket_matches:
        xu_ly_sau2.extend(bracket_matches)
        for bm in bracket_matches:
            job = job.replace(f"[{bm}]", "")

    # 2. Tách số kèm "up to", "$", "M", "k" hoặc ~ sang ky_tu
    number_pattern = r'(?i)(up to\s*)?(\$)?(\d+\.?\d*)([kM\$~])?'
    number_matches = re.findall(number_pattern, job)
    if number_matches:
        for match in number_matches:
            full_match = ''.join(match).strip()
            if full_match:
                ky_tu.append(full_match)
                job = job.replace(full_match, '')

    # 3. Lấy dấu / nếu ở đầu chuỗi
    if job.startswith("/"):
        ky_tu.append("/")
        job = job[1:]

    # 4. Xóa dấu . hoặc / ở đầu chuỗi **trừ khi là .NET**
    if job.startswith(('.', '/')) and not job.lower().startswith('.net'):
        job = job.lstrip('./')

    # 5. Làm sạch khoảng trắng thừa
    job_clean2 = job.strip()

    ky_tu_str = " ".join(ky_tu) if ky_tu else None
    xu_ly_sau2_str = ", ".join(xu_ly_sau2) if xu_ly_sau2 else None

    return pd.Series([ky_tu_str, xu_ly_sau2_str, job_clean2])

# Áp dụng cho job_names_clean
df[['ky_tu', 'xu_ly_sau2', 'job_names_clean2']] = df['job_names_clean'].apply(clean_job_name_step4)

# Giữ nguyên index gốc
df = df.sort_index()


In [10]:
# Xóa cột job_names, job_names_clean vì không cần nữa
df = df.drop(['job_names', 'job_names_clean'], axis=1)

In [11]:
# đổi tên cột job_names sạch lại thành job_names
df = df.rename(columns={'job_names_clean2': 'job_names'})

# xử lý phân loại cho job_names

Được, tôi sẽ phân tích chi tiết cho bạn cách **gom các tên job khác nhau về nhóm công việc chung**, từ ý tưởng đến cách làm, ưu/nhược điểm, và cách đánh giá hiệu quả.

---

### **1. Mục tiêu**

Bạn có cột `job_names` với nhiều tên công việc khác nhau, nhưng thực tế nhiều tên chỉ khác nhau về chi tiết (ví dụ: `Java Developer`, `Java Backend Developer`, `Backend Java Developer`). Bạn muốn gom chúng về **nhóm chính**, ví dụ:

* Backend Developer
* Frontend Developer
* Fullstack Developer
* Data Scientist / AI Engineer
* QA / Tester
* Project Manager / Product Owner

---

### **2. Ý tưởng lọc và gom nhóm**

**Cơ sở để phân nhóm chính xác:**

1. **Keyword-based (dựa vào từ khóa)**

   * Xác định các từ khóa đặc trưng của nhóm công việc:

     * Backend: `backend`, `java`, `python`, `golang`, `c++`, `nodejs`
     * Frontend: `frontend`, `react`, `vue`, `angular`, `ui/ux`
     * Fullstack: `fullstack`
     * QA / Tester: `qa`, `tester`, `automation`, `manual`
     * AI / Data: `data`, `ai`, `machine learning`, `ml`, `data scientist`, `analyst`
     * DevOps / Cloud: `devops`, `cloud`, `sre`
     * Manager / Lead: `manager`, `lead`, `head`, `director`

   * Đây là **cách phổ biến và dễ làm**, cho hiệu quả cao nếu tập trung vào từ khóa quan trọng.

2. **Regex hoặc text preprocessing**

   * Chuyển tất cả về lowercase
   * Xóa các ký tự thừa, dấu câu, khoảng trắng
   * Tách các từ thành danh sách, kiểm tra sự xuất hiện của từ khóa

3. **Mapping manual + semi-automated**

   * Tạo dictionary mapping các job phổ biến → nhóm chính:

     ```python
     mapping = {
         'java': 'Backend Developer',
         'python': 'Backend Developer',
         'fullstack': 'Fullstack Developer',
         'frontend': 'Frontend Developer',
         'qa': 'QA / Tester',
         'data': 'Data / AI',
         'ai': 'Data / AI',
         'devops': 'DevOps / Cloud',
         'manager': 'Manager / Lead'
     }
     ```
   * Dùng thuật toán kiểm tra từng tên job và gán vào nhóm dựa trên từ khóa đầu tiên match.

4. **Text embedding / NLP (nâng cao)**

   * Nếu bạn muốn chính xác ~90%+, có thể dùng embedding (BERT, SentenceTransformer) để đo similarity giữa tên job và nhóm mẫu.
   * Ví dụ: tính cosine similarity giữa `Java Backend Developer` và nhóm mẫu `Backend Developer`.

---

### **3. Cách làm thực tế**

Một workflow điển hình:

1. **Tiền xử lý dữ liệu**

   ```python
   df['job_clean'] = df['job_names'].str.lower().str.replace(r'[^a-z0-9\s]', '', regex=True)
   ```

2. **Xây từ khóa cho nhóm**

   ```python
   job_groups = {
       'Backend Developer': ['backend', 'java', 'python', 'golang', 'c++', 'nodejs'],
       'Frontend Developer': ['frontend', 'react', 'vue', 'angular', 'ui ux'],
       'Fullstack Developer': ['fullstack'],
       'QA / Tester': ['qa', 'tester', 'automation', 'manual'],
       'Data / AI': ['data', 'ai', 'machine learning', 'ml', 'analyst', 'engineer'],
       'DevOps / Cloud': ['devops', 'cloud', 'sre'],
       'Manager / Lead': ['manager', 'lead', 'head', 'director']
   }
   ```

3. **Gom nhóm theo keyword**

   ```python
   def assign_group(job):
       for group, keywords in job_groups.items():
           if any(k in job for k in keywords):
               return group
       return 'Other'

   df['job_group'] = df['job_clean'].apply(assign_group)
   ```

4. **Kiểm tra kết quả**

   ```python
   df[['job_names', 'job_group']].head(20)
   ```

---

### **4. Thuận lợi / Bất lợi**

**Thuận lợi:**

* Dễ triển khai, không cần AI quá phức tạp.
* Có thể xử lý hàng nghìn job nhanh chóng.
* Tinh chỉnh keyword hoặc mapping nếu thấy nhiều lỗi.

**Bất lợi:**

* Nếu tên job quá dài, lộn xộn, nhiều ký tự thừa → đôi khi gán nhầm nhóm.
* Các job mới, lạ không có keyword → vào nhóm “Other”.
* Không nhận biết ngữ cảnh: `Java` đôi khi là Backend nhưng trong job đặc biệt có thể không phải.

---

### **5. Đánh giá hiệu quả**

Để đảm bảo **độ chính xác ~90%**:

1. Lấy **sample 100-200 job** và gán nhóm thủ công → so sánh với kết quả tự động.
2. Tính **accuracy**:
   [
   \text{Accuracy} = \frac{\text{Số job đúng nhóm}}{\text{Tổng số job sample}} \times 100%
   ]
3. Tinh chỉnh **từ khóa** cho các nhóm bị sai.
4. Lặp lại vài lần cho đến khi độ chính xác >90%.

---

Nếu bạn muốn, tôi có thể viết cho bạn **một đoạn code Python đầy đủ**, từ tiền xử lý → gán nhóm → kiểm tra hiệu quả ~90% cho dữ liệu job names bạn gửi.

Bạn có muốn tôi làm luôn không?


In [12]:
df['job_clean'] = df['job_names'].str.lower().str.replace(r'[^a-z0-9\s]', '', regex=True)

# 2. Định nghĩa các nhóm và từ khóa, ưu tiên theo thứ tự quan trọng
job_groups = {
    'Manager / Lead': ['manager', 'lead', 'head', 'director', 'owner', 'pm', 'project manager', 'product owner', 'deputy', 'erp manager'],
    'DevOps / Cloud': ['devops', 'cloud', 'sre', 'infrastructure', 'system engineer', 'database reliability'],
    'Data / AI': ['data', 'ai', 'machine learning', 'ml', 'data scientist', 'analyst', 'vision', 'ai engineer'],
    'Mobile Developer': ['ios', 'android', 'flutter', 'mobile'],
    'Fullstack Developer': ['fullstack', 'full stack', 'fullstack engineer', 'full stack engineer'],
    'Backend Developer': ['back end','backend', 'java', 'python', 'golang', 'c++', 'nodejs', 'c#', '.net', 'odoo', 'firmware'],
    'Frontend Developer': ['frontend', 'react', 'vue', 'angular', 'ui ux', 'ui/ux', 'reactjs', 'nextjs', 'unity'],
    'QA / Tester': ['kiểm thử','Kiểm thử','qa', 'tester', 'automation', 'manual', 'qc'],
    'Other': []
}

# 3. Hàm gán nhóm
def assign_group(job):
    for group in ['Manager / Lead', 'DevOps / Cloud', 'Data / AI', 'Mobile Developer', 
                  'Fullstack Developer', 'Backend Developer', 'Frontend Developer', 'QA / Tester', 'Other']:
        keywords = job_groups[group]
        if any(k in job for k in keywords):
            return group
    return 'Other'

df['job_group'] = df['job_clean'].apply(assign_group)

In [13]:
# kiểm tra chênh lệch giữa các nhóm
df.job_group.value_counts()

job_group
Other                  234
Data / AI              169
Backend Developer      160
Manager / Lead          91
QA / Tester             87
Fullstack Developer     79
Frontend Developer      64
DevOps / Cloud          54
Mobile Developer        40
Name: count, dtype: int64

# tiếp tục xử lý các cột phát sinh từ cột làm sạch ban đầu để tiếp tục lấy thông tin

In [18]:
df

,company_names,salaries,position_names,kind_jobs,array_skills,locate_names,exp_skills,domain_arr,post_dates_formatted,level,xu_ly_sau,ky_tu,xu_ly_sau2,job_names,job_group
0,SHINHAN DS,Up to 40m vnd,Backend Developer,At office,"['Java', 'API', 'Spring Boot', 'Database', 'PL...","Floor 3, The Mett Tower, 15 Tran Bach Dang Str...",Job Requirements:\nBachelor’s degree in Comput...,"['Banking', 'Financial Services']",25/11/2025 15:05,senior,NaN,None,None,Java Backend Developer,Backend Developer
1,MEGAZONE,You'll love it,Manual Tester,At office,"['Tester', 'Automation Test', 'MLOps', 'AI', '...","54 Lieu Giai, Ba Dinh, Ha Noi",Basic Requirements\nMinimum 4 years of experie...,['IT Services and IT Consulting'],25/11/2025 15:05,NaN,"QA QC, AI Model Management Platform",None,None,Application Tester,QA / Tester
2,ELCA,You'll love it,AI / Machine Learning Engineer,Hybrid (flexible between home and office),"['AI', 'LLM', 'Deep Learning', 'Machine Learni...","CII Building, 152 Dien Bien Phu, Thanh My Tay ...","Bachelor’s degree in computer science, AI, or ...","['IT Services and IT Consulting', 'Software De...",25/11/2025 14:05,senior,NaN,None,None,AI Engineer,Data / AI
3,SenPrints,You'll love it,Fullstack Developer,At office,"['Mockups', 'UI-UX', 'Figma', 'API', 'Python',...","T3-OF08, Timescity, 458 Minh Khai, Hai Ba Trun...",Kinh nghiệm Fullstack từ 1–3 năm (hoặc tương đ...,['Publishing and Printing'],25/11/2025 14:05,NaN,"Python, NodeJS, API",None,None,Fullstack Developer Render Mockup,Fullstack Developer
4,LG CNS Việt Nam,You'll love it,Cloud Engineer,At office,"['DevOps', 'Cloud']","Tầng 15, tòa Keangnam Landmark 72, Mễ Trì, Nam...","Bachelor’s degree or higher in IT, Computer Sc...",['AI Software & Services'],25/11/2025 13:05,NaN,NaN,None,None,Cloud/DevOps Engineer,DevOps / Cloud
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
988,Golden Gate,"500 - 1,000 USD",Data Scientist,At office,"['Data Science', 'Data Analysis', 'Big Data', ...","315 Trường Chinh, Thanh Xuan, Ha Noi",Bachelor‘s degree in a quantitative field (e.g...,"['Food and Beverage', 'Tourism and Hospitality...",26/11/2025 20:51,senior,"Strong Python, SQL",None,None,Data Scientist,Data / AI
989,TymeX,You'll love it,Backend Developer,Hybrid (flexible between home and office),"['Java', 'Spring', 'Microservices', 'Python', ...","Capital Place, 29 Lieu Giai, Ngoc Khanh Ward, ...",Requirements\nYou will be a perfect match if y...,['Financial Services'],26/11/2025 20:51,mid,"Microservice, AWS",/,None,Java Engineer Hanoi,Backend Developer
990,Trung tâm CNTT Tập đoàn Bảo Việt,You'll love it,Database Administrator,At office,"['Database', 'Microsoft SQL Server', 'PostgreS...","71 Ngô Sĩ Liên, Dong Da, Ha Noi","1. Yêu cầu tối thiểu:\n- Trình độ, chuyên môn:...","['Insurance', 'Banking', 'Software Products an...",26/11/2025 20:51,NaN,NaN,None,None,CV Quản trị cơ sở dữ liệu Database Administrator,Data / AI
991,NAVER VIETNAM,You'll love it,Mobile Application Developer,Hybrid (flexible between home and office),"['Swift', 'Mobile Apps', 'Objective C', 'iOS']","4th and 9th Floor, Block B, Viettel Building, ...",Your qualifications: \nBackground in computer ...,"['IT Hardware and Computing', 'IT Services and...",26/11/2025 20:51,NaN,NaN,None,None,iOS Engineer Dictionary,Mobile Developer


In [15]:
# Xóa cột job_clean vì không cần nữa
df = df.drop([ 'job_clean'], axis=1)